raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
+            // to validate the epoch from the response against our current state.
+            maybeBecomeFollower(leaderId, epoch);
+            return Optional.of(true);
+        } else if (error == Errors.BROKER_NOT_AVAILABLE) {
+            if (quorum.isObserver()) {
+                becomeUnattachedFollower(epoch);
+                return Optional.of(true);
+            } else {
+                return Optional.of(false);
             }
-
-            responseEpoch = fetchRecordsResponse.leaderEpoch();
-            responseLeaderId = optionalLeaderId(fetchRecordsResponse.leaderId());
-        } else if (data instanceof VoteResponseData) {
-            VoteResponseData voteResponse = (VoteResponseData) data;
-            responseError = Errors.forCode(voteResponse.errorCode());
-            responseEpoch = voteResponse.leaderEpoch();
-            responseLeaderId = optionalLeaderId(voteResponse.leaderId());
-        } else if (data instanceof BeginQuorumEpochResponseData) {
-            BeginQuorumEpochResponseData beginEpochResponse = (BeginQuorumEpochResponseData) data;
-            responseError = Errors.forCode(beginEpochResponse.errorCode());
-            responseEpoch = beginEpochResponse.leaderEpoch();
-            responseLeaderId = optionalLeaderId(beginEpochResponse.leaderId());
-        } else if (data instanceof EndQuorumEpochResponseData) {
-            EndQuorumEpochResponseData endEpochResponse = (EndQuorumEpochResponseData) data;
-            responseError = Errors.forCode(endEpochResponse.errorCode());
-            responseEpoch = endEpochResponse.leaderEpoch();
-            responseLeaderId = optionalLeaderId(endEpochResponse.leaderId());
-        } else if (data instanceof FindQuorumResponseData) {
-            FindQuorumResponseData findQuorumResponse = (FindQuorumResponseData) data;
-            responseError = Errors.forCode(findQuorumResponse.errorCode());
-            responseEpoch = findQuorumResponse.leaderEpoch();
-            responseLeaderId = optionalLeaderId(findQuorumResponse.leaderId());
-
-            // TODO: Find a new home for this. The main point is that we need to do it even
-            // if the response has an error
-            updateVoterConnections(findQuorumResponse);
-        } else {
-            throw new IllegalStateException("Received unexpected response " + response);
+        } else if (error == Errors.INCONSISTENT_GROUP_PROTOCOL) {
+            // For now we treat this as a fatal error. Once we have support for quorum
+            // reassignment, this error could suggest that either we or the recipient of
+            // the request just has stale voter information, which means we can retry
+            // after backing off.
+            throw new IllegalStateException("Received error indicating inconsistent voter sets");
+        } else if (error == Errors.INVALID_REQUEST) {
+            throw new IllegalStateException("Received unexpected invalid request error");
```

Comment:
Should we pass the error here as well?

Reply:
Not sure what the benefit would be. The message already indicates it was an invalid request error.

Reply:
Oh, you are right.